src/zkp/generator.rs

```diff
@@ -1,7 +1,79 @@
 use core::{fmt, str};
 use ffi;
-use {constants, from_hex, Error, Secp256k1, SecretKey, Signing, Tag};
+#[cfg(feature = "rand")]
+use rand::Rng;
+use {constants, from_hex, Error, Secp256k1, Signing, Tag};
+/// Represents a blinding factor/Tweak on secp256k1 curve
+///
+/// This is same as [`crate::SecretKey`], but is allowed to take
+/// zero value and values outside of secp curve order
+pub struct Tweak([u8; constants::SECRET_KEY_SIZE]);
+impl_array_newtype!(Tweak, u8, constants::SECRET_KEY_SIZE);
+
+/// The number 1 encoded as a secret key
+pub const ZERO_TWEAK: Tweak = Tweak([
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+]);
+
+impl fmt::Debug for Tweak {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Tweak(")?;
+        for i in self[..].iter().cloned() {
+            write!(f, "{:02x}", i)?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl fmt::LowerHex for Tweak {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for ch in &self.0[..] {
+            write!(f, "{:02x}", *ch)?;
+        }
+        Ok(())
+    }
+}
 
+impl fmt::Display for Tweak {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl str::FromStr for Tweak {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Tweak, Error> {
+        let mut res = [0; constants::SECRET_KEY_SIZE];
+        match from_hex(s, &mut res) {
+            Ok(constants::SECRET_KEY_SIZE) => Ok(Tweak(res)),
+            _ => Err(Error::InvalidTweak),
+        }
+    }
+}
+
+impl Tweak {
+    /// Generate a new random Tweak
+    #[cfg(feature = "rand")]
+    pub fn new<R: Rng + ?Sized>(rng: &mut R) -> Tweak {
+        let mut ret = [0u8; 32];
```

Comment:
Probably best to use the constant everywhere.

```suggestion
        let mut ret = [0u8; constants::SECRET_KEY_SIZE];
```